File: projects/rpn2tex/migrations_archive/go-module-by-module-1/token_test.go
package rpn2tex

import "testing"

func TestTokenType_String(t *testing.T) {
	tests := []struct {
		name string
		tt   TokenType
		want string
	}{
		{"NUMBER", NUMBER, "NUMBER"},
		{"PLUS", PLUS, "PLUS"},
		{"MINUS", MINUS, "MINUS"},
		{"MULT", MULT, "MULT"},
		{"DIV", DIV, "DIV"},
		{"EOF", EOF, "EOF"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.tt.String(); got != tt.want {
				t.Errorf("TokenType.String() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestToken_String(t *testing.T) {
	tests := []struct {
		name  string
		token Token
		want  string
	}{
		{
			name:  "NUMBER token",
			token: Token{Type: NUMBER, Value: "42", Line: 1, Column: 1},
			want:  `Token(NUMBER, "42", 1, 1)`,
		},
		{
			name:  "PLUS token",
			token: Token{Type: PLUS, Value: "+", Line: 1, Column: 5},
			want:  `Token(PLUS, "+", 1, 5)`,
		},
		{
			name:  "Decimal number",
			token: Token{Type: NUMBER, Value: "3.14", Line: 2, Column: 10},
			want:  `Token(NUMBER, "3.14", 2, 10)`,
		},
		{
			name:  "EOF token",
			token: Token{Type: EOF, Value: "", Line: 1, Column: 10},
			want:  `Token(EOF, "", 1, 10)`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.token.String(); got != tt.want {
				t.Errorf("Token.String() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestToken_Creation(t *testing.T) {
	// Test that Token struct can be created with all fields
	token := Token{
		Type:   MINUS,
		Value:  "-",
		Line:   3,
		Column: 7,
	}

	if token.Type != MINUS {
		t.Errorf("Token.Type = %v, want MINUS", token.Type)
	}
	if token.Value != "-" {
		t.Errorf("Token.Value = %v, want '-'", token.Value)
	}
	if token.Line != 3 {
		t.Errorf("Token.Line = %v, want 3", token.Line)
	}
	if token.Column != 7 {
		t.Errorf("Token.Column = %v, want 7", token.Column)
	}
}

func TestTokenType_Values(t *testing.T) {
	// Verify that token types have distinct values
	types := []TokenType{NUMBER, PLUS, MINUS, MULT, DIV, EOF}
	seen := make(map[TokenType]bool)

	for _, tt := range types {
		if seen[tt] {
			t.Errorf("Duplicate TokenType value: %v", tt)
		}
		seen[tt] = true
	}

	// Verify expected ordering (iota increments from 0)
	if NUMBER != 0 {
		t.Errorf("NUMBER = %d, want 0", NUMBER)
	}
	if EOF != 5 {
		t.Errorf("EOF = %d, want 5", EOF)
	}
}

File: projects/rpn2tex/migrations/java-module-by-module-3/src/test/java/com/rpn2tex/TokenTest.java
package com.rpn2tex;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for the Token record.
 */
class TokenTest {

    @Test
    void testTokenCreation() {
        // Test basic token creation
        Token token = new Token(TokenType.NUMBER, "5", 1, 1);

        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("5", token.value());
        assertEquals(1, token.line());
        assertEquals(1, token.column());
    }

    @Test
    void testTokenEquality() {
        // Two tokens with same fields should be equal
        Token token1 = new Token(TokenType.NUMBER, "5", 1, 1);
        Token token2 = new Token(TokenType.NUMBER, "5", 1, 1);

        assertEquals(token1, token2);
        assertEquals(token1.hashCode(), token2.hashCode());
    }

    @Test
    void testTokenInequality() {
        // Tokens with different fields should not be equal
        Token token1 = new Token(TokenType.NUMBER, "5", 1, 1);
        Token token2 = new Token(TokenType.NUMBER, "3", 1, 1);
        Token token3 = new Token(TokenType.PLUS, "5", 1, 1);
        Token token4 = new Token(TokenType.NUMBER, "5", 2, 1);
        Token token5 = new Token(TokenType.NUMBER, "5", 1, 2);

        assertNotEquals(token1, token2, "Different values should not be equal");
        assertNotEquals(token1, token3, "Different types should not be equal");
        assertNotEquals(token1, token4, "Different lines should not be equal");
        assertNotEquals(token1, token5, "Different columns should not be equal");
    }

    @Test
    void testTokenStringRepresentation() {
        // Test toString format: Token(TYPE, "value", line:column)
        Token numberToken = new Token(TokenType.NUMBER, "5", 1, 1);
        assertEquals("Token(NUMBER, \"5\", 1:1)", numberToken.toString());

        Token plusToken = new Token(TokenType.PLUS, "+", 1, 3);
        assertEquals("Token(PLUS, \"+\", 1:3)", plusToken.toString());

        Token eofToken = new Token(TokenType.EOF, "", 1, 10);
        assertEquals("Token(EOF, \"\", 1:10)", eofToken.toString());
    }

    @ParameterizedTest
    @CsvSource({
        "NUMBER, '5', 1, 1",
        "NUMBER, '3.14', 1, 5",
        "NUMBER, '-2', 2, 1",
        "PLUS, '+', 1, 3",
        "MINUS, '-', 1, 3",
        "MULT, '*', 1, 5",
        "DIV, '/', 1, 7",
        "EOF, '', 1, 10"
    })
    void testVariousTokenTypes(String typeStr, String value, int line, int column) {
        // Test creating tokens of different types
        TokenType type = TokenType.valueOf(typeStr);
        Token token = new Token(type, value, line, column);

        assertEquals(type, token.type());
        assertEquals(value, token.value());
        assertEquals(line, token.line());
        assertEquals(column, token.column());
    }

    @Test
    void testNullTypeThrowsException() {
        // Null type should throw NullPointerException
        assertThrows(NullPointerException.class, () -> {
            new Token(null, "5", 1, 1);
        });
    }

    @Test
    void testNullValueThrowsException() {
        // Null value should throw NullPointerException
        assertThrows(NullPointerException.class, () -> {
            new Token(TokenType.NUMBER, null, 1, 1);
        });
    }

    @Test
    void testTokenWithDecimalValue() {
        // Test token with decimal number
        Token token = new Token(TokenType.NUMBER, "3.14", 1, 1);

        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("3.14", token.value());
    }

    @Test
    void testTokenWithNegativeValue() {
        // Test token with negative number
        Token token = new Token(TokenType.NUMBER, "-5", 1, 1);

        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("-5", token.value());
    }

    @Test
    void testEOFToken() {
        // Test EOF token with empty value
        Token token = new Token(TokenType.EOF, "", 1, 10);

        assertEquals(TokenType.EOF, token.type());
        assertEquals("", token.value());
        assertTrue(token.value().isEmpty());
    }

    @Test
    void testTokenPositionTracking() {
        // Test that position is tracked correctly across different locations
        Token token1 = new Token(TokenType.NUMBER, "5", 1, 1);
        Token token2 = new Token(TokenType.PLUS, "+", 1, 3);
        Token token3 = new Token(TokenType.NUMBER, "3", 2, 1);

        assertEquals(1, token1.line());
        assertEquals(1, token1.column());

        assertEquals(1, token2.line());
        assertEquals(3, token2.column());

        assertEquals(2, token3.line());
        assertEquals(1, token3.column());
    }

    @Test
    void testTokenImmutability() {
        // Verify that Token is immutable (record automatically provides this)
        Token token = new Token(TokenType.NUMBER, "5", 1, 1);

        // Store original values
        TokenType originalType = token.type();
        String originalValue = token.value();
        int originalLine = token.line();
        int originalColumn = token.column();

        // Create a "modified" token - must create new instance
        Token modifiedToken = new Token(TokenType.PLUS, "6", 2, 2);

        // Original token should be unchanged
        assertEquals(originalType, token.type());
        assertEquals(originalValue, token.value());
        assertEquals(originalLine, token.line());
        assertEquals(originalColumn, token.column());
    }

    @Test
    void testTokenRecordComponentAccessors() {
        // Test that record component accessors work correctly
        Token token = new Token(TokenType.MULT, "*", 3, 5);

        // All accessor methods should be available and work correctly
        assertNotNull(token.type());
        assertNotNull(token.value());
        assertTrue(token.line() > 0);
        assertTrue(token.column() > 0);
    }
}
